ls1intum/Artemis
# Integrate feedback validation service. Adapt automatic text feedback suggestion flow

### Checklist

### Motivation and Context
Currently the automatic text feedback system deals with a high number of "false positives" - feedback suggestions that had to be adapted by a tutor. This PR improves this by integrating a feedback validation service and considering the entire text cluster in the process of creating automatic feedback. 

### Description
Features: 

- Integrates the text feedback validation service which calculates a "confidence" based 
   on a (use-case adapted) BLEU algorithm https://en.wikipedia.org/wiki/BLEU .

- Implements a normalization approach for score generation towards the visionary scenario of "1 text cluster deserving 1 feedback". 

- Adapts the nearest neighbor search for comment generation.

- Adapts the flow leading of suggested automatic feedback being presented to tutors. 

Refactors: 

- Extracts the suggestFeedback method 

- Extracts the textClusterUtilityService class providing methods to work with the feedback of a text cluster

Tests: 

- Adds test suite for automatic feedback service.

- Adds test suite for text cluster utility service.

- Adds test suite for text feedback validation connector.


### Steps for Testing

0. Run container from https://github.com/ls1intum/Athene/pull/4 
1. Log in to Artemis
2. Navigate to Course Administration
3. Create a text exercise with automatic feedback enabled
4. Create 10+ similar text submissions to said exercise (English, full sentences, you can reuse the same submissions) 
5. For each submission, click on 'create text blocks automatically'
6. For each submission, start giving feedback on the text blocks
7. Once you have given enough (min. 2) similar feedback on similar text blocks, a feedback suggestion should pop up.
